lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -3125,28 +3117,44 @@ joinStakePool ctx currentEpoch knownPools pid poolStatus wid =
     nl = ctx ^. networkLayer
 
 -- | Helper function to factor necessary logic for quitting a stake pool.
-quitStakePool
-    :: forall ctx s k
-     . HasDBLayer IO s k ctx
-    => ctx
+validatedQuitStakePoolAction
+    :: forall s k
+     . DBLayer IO s k
     -> WalletId
     -> Withdrawal
-    -> ExceptT ErrStakePoolDelegation IO DelegationAction
-quitStakePool ctx wid wdrl = db & \DBLayer{..} -> do
-    walMeta <- mapExceptT atomically
-        $ withExceptT ErrStakePoolDelegationNoSuchWallet
-        $ withNoSuchWallet wid
-        $ readWalletMeta wid
-
-    rewards <- liftIO
-        $ fetchRewardBalance @ctx @s @k ctx wid
+    -> IO DelegationAction
+validatedQuitStakePoolAction db@DBLayer{..} walletId withdrawal = do
+    walletMeta <- atomically (readWalletMeta walletId)
+        >>= maybe
+            (throw (ExceptionStakePoolDelegation
+                (ErrStakePoolDelegationNoSuchWallet
+                    (ErrNoSuchWallet walletId))))
+            pure
+    rewards <- liftIO $ fetchRewardBalance @s @k db walletId
+    Quit <$
+        either (throw . ExceptionStakePoolDelegation . ErrStakePoolQuit) pure
+            (guardQuit (walletMeta ^. #delegation) withdrawal rewards)
```

Comment:
IMO:
```haskell
guard
pure Quit
```
is simpler to understand than

```haskell
Quit <$
    guard
```

Reply:
[`Control.Exception.throw`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Control-Exception.html#g:2): 
> WARNING: You may want to use throwIO instead so that your pure code stays exception-free.

Reply:
Good points, fixed.

Reply:
Ok, changed.